File: backend/src/safebrowser/SafeBrowserNavigationPolicy.ts
import { BlockDecisionStateService } from './BlockDecisionStateStore.js';
import type { BlockDecisionState, OpaqueFamilyId, OpaqueProfileId } from './types.js';
import { canonicalizeDomain } from '../web/canonicalize.js';
import { resolveEffectiveSafeSearchMode } from '../web/policy.js';
import type { WebFilterEngine } from '../web/WebFilterEngine.js';
import type { ClassifierResult, SafeSearchDirective, SafeSearchMode } from '../web/types.js';
import type { SupportedLocale } from '../i18n/types.js';

export type NavigationOutcome =
  | { status: 'ALLOW'; safeSearchMode: SafeSearchMode }
  | { status: 'BLOCK' | 'REVIEW'; blockDecision: BlockDecisionState };

export type NavigationErrorCode = 'INVALID_URL';

export class NavigationPolicyError extends Error {
  readonly code: NavigationErrorCode;
  constructor(code: NavigationErrorCode) {
    super('URL could not be canonicalized to a domain.');
    this.name = 'NavigationPolicyError';
    this.code = code;
  }
}

/**
 * doc 14's Safe-Browser-specific entry point: runs the shared web/
 * decision pipeline (rules -> optional classifier), applies the doc 14
 * layer-1 SafeSearch directive when the destination search service
 * supports it, and -- only for a BLOCK/REVIEW outcome -- records a local
 * BlockDecisionState carrying the full URL/title this surface alone is
 * permitted to retain (doc 14's visibility matrix). An ALLOW outcome is
 * never persisted here: doc 15 "PCA services do not persist readable
 * app-use timelines" extends to this module -- an allowed page is not
 * itself a retained event, only a block/review decision is.
 *
 * PCA-16A correction (BACKEND_I18N_NOT_WIRED): `locale` is threaded through
 * to WebFilterEngine.decide() from HERE -- the real production boundary
 * where a navigation request is evaluated on behalf of a specific child
 * profile/device, which is where the presentation locale is actually known
 * (the caller resolves it from the requesting device/profile's language
 * preference; this class never guesses or defaults it beyond the explicit
 * fallback below). The resulting `blockDecision.reasonCode` a parent/child
 * sees is genuinely localized end-to-end, not just in an unused helper.
 */
export class SafeBrowserNavigationPolicy {
  private readonly engine: WebFilterEngine;
  private readonly blockDecisions: BlockDecisionStateService;

  constructor(engine: WebFilterEngine, blockDecisions: BlockDecisionStateService) {
    this.engine = engine;
    this.blockDecisions = blockDecisions;
  }

  async evaluateNavigation(
    familyId: OpaqueFamilyId,
    profileId: OpaqueProfileId,
    url: string,
    options: { pageTitle?: string | null; classifierResult?: ClassifierResult; safeSearch?: SafeSearchDirective } = {},
    locale: SupportedLocale = 'en',
  ): Promise<NavigationOutcome> {
    const domain = canonicalizeDomain(url);
    if (domain === null) throw new NavigationPolicyError('INVALID_URL');

    const decision = await this.engine.decide(familyId, domain, { classifierResult: options.classifierResult }, locale);

    if (decision.outcome === 'ALLOW') {
      const safeSearchMode = options.safeSearch ? resolveEffectiveSafeSearchMode(options.safeSearch) : 'OFF';
      return { status: 'ALLOW', safeSearchMode };
    }

    const blockDecision = await this.blockDecisions.record(
      familyId,
      profileId,
      url,
      options.pageTitle ?? null,
      decision,
    );
    return { status: decision.outcome === 'REVIEW' ? 'REVIEW' : 'BLOCK', blockDecision };
  }
}

File: android/app/src/main/java/org/pca/app/feature/webprotection/safebrowser/SafeBrowserNavigationPolicy.kt
package org.pca.app.feature.webprotection.safebrowser

import java.util.UUID
import org.pca.app.feature.webprotection.engine.WebFilterEngine
import org.pca.app.feature.webprotection.policy.ClassifierResult
import org.pca.app.feature.webprotection.policy.OpaqueFamilyId
import org.pca.app.feature.webprotection.policy.SafeSearchDirective
import org.pca.app.feature.webprotection.policy.SafeSearchMode
import org.pca.app.feature.webprotection.policy.VpnMetadataDecision
import org.pca.app.feature.webprotection.policy.WebDecisionOutcome
import org.pca.app.feature.webprotection.policy.WebLocale
import org.pca.app.feature.webprotection.policy.canonicalizeDomain
import org.pca.app.feature.webprotection.policy.isBlockRequestable
import org.pca.app.feature.webprotection.policy.resolveEffectiveSafeSearchMode
import org.pca.app.persistence.entity.WebVisitAction
import org.pca.app.persistence.repository.WebVisitRepository

/**
 * doc 14's Safe-Browser-specific ANDROID RUNTIME entry point -- this is the
 * seam the actual PCA Safe Browser WebView surface calls on every navigation
 * attempt (`shouldOverrideUrlLoading`/`shouldInterceptRequest`). Mirrors
 * `backend/src/safebrowser/SafeBrowserNavigationPolicy.ts`'s pipeline
 * exactly: canonicalize -> [WebFilterEngine.decide] -> ALLOW (apply
 * SafeSearch) or BLOCK/REVIEW (persist a [BlockDecisionState]).
 *
 * Persistence: an ALLOW outcome is NEVER persisted here (doc 15: "PCA
 * services do not persist readable app-use timelines" -- an allowed page is
 * not itself a retained event). A BLOCK/REVIEW outcome is persisted via the
 * EXISTING, already-encrypted [WebVisitRepository]/`WebVisitEntity` (doc 10
 * Section 4.2 already approves exactly this URL/title persistence, scoped to
 * the Safe Browser browsing context only -- this class is the only caller in
 * this lane that ever supplies a full `url`/`pageTitle`, satisfying the
 * "Safe Browser context may see a full URL only within its own legitimate
 * browsing context" hard requirement structurally: nothing outside
 * `safebrowser/` in this package ever touches a full URL string).
 */
class SafeBrowserNavigationPolicy(
    private val engine: WebFilterEngine,
    private val webVisits: WebVisitRepository,
    private val now: () -> Long = { System.currentTimeMillis() },
    private val idGenerator: () -> String = { UUID.randomUUID().toString() },
    private val enrollmentSafeSearchDefault: () -> SafeSearchDirective? = { null },
) {
    suspend fun evaluateNavigation(
        familyId: OpaqueFamilyId,
        profileId: OpaqueProfileId,
        deviceId: String,
        url: String,
        pageTitle: String? = null,
        classifierResult: ClassifierResult? = null,
        vpnDecision: VpnMetadataDecision? = null,
        safeSearch: SafeSearchDirective? = null,
        locale: WebLocale = WebLocale.EN,
    ): NavigationOutcome {
        val domain = canonicalizeDomain(url) ?: throw NavigationError.InvalidUrl

        val decision = engine.decide(familyId, domain, classifierResult, vpnDecision, locale)

        if (decision.outcome == WebDecisionOutcome.ALLOW) {
            val effectiveSafeSearch = mergeEnrollmentMinimum(enrollmentSafeSearchDefault(), safeSearch)
            val safeSearchMode = effectiveSafeSearch?.let(::resolveEffectiveSafeSearchMode) ?: SafeSearchMode.OFF
            val rewrite = applySafeSearchQueryParameter(url, domain, safeSearchMode)
            return NavigationOutcome.Allow(safeSearchMode, rewrite.url, rewrite.applied)
        }

        val id = idGenerator()
        val timestamp = now()
        val requestable = decision.outcome != WebDecisionOutcome.ALLOW && isBlockRequestable(decision.source)

        webVisits.record(
            deviceId = deviceId,
            domain = domain,
            url = url,
            title = pageTitle,
            classificationCategory = decision.source.name,
            ruleOrModelVersion = decision.reasonId.name,
            action = if (decision.outcome == WebDecisionOutcome.REVIEW) WebVisitAction.REVIEW else WebVisitAction.BLOCKED,
            timestampEpochMillis = timestamp,
            id = id,
        )

        val blockDecision = BlockDecisionState(
            id = id,
            familyId = familyId,
            profileId = profileId,
            domain = domain,
            url = url,
            pageTitle = pageTitle,
            outcome = decision.outcome,
            source = decision.source,
            reasonId = decision.reasonId,
            reasonCode = decision.reasonCode,
            requestable = requestable,
            createdAtEpochMillis = timestamp,
        )
        return NavigationOutcome.Held(blockDecision)
    }

    private fun mergeEnrollmentMinimum(
        enrollmentMinimum: SafeSearchDirective?,
        requested: SafeSearchDirective?,
    ): SafeSearchDirective? {
        if (requested == null) return enrollmentMinimum
        if (!requested.serviceSupportsSafeSearch || enrollmentMinimum == null) return requested
        val mode = if (enrollmentMinimum.mode.rank() > requested.mode.rank()) enrollmentMinimum.mode else requested.mode
        return SafeSearchDirective(mode, serviceSupportsSafeSearch = true)
    }

    private fun SafeSearchMode.rank(): Int = when (this) {
        SafeSearchMode.OFF -> 0
        SafeSearchMode.MODERATE -> 1
        SafeSearchMode.STRICT -> 2
    }
}
